cocotb/bus.py

```diff
@@ -93,19 +91,18 @@ def __init__(self, entity, name, signals, optional_signals=[], bus_separator="_"
             else:
                 signame = sig_name
 
-            if array_idx is not None:
-                signame += "[{:d}]".format(array_idx)
-
-            self._entity._log.debug("Signal name {}".format(signame))
             if hasattr(entity, signame):
-                self._add_signal(attr_name, signame)
+                self._add_signal(attr_name, signame, array_idx)
             else:
                 self._entity._log.debug("Ignoring optional missing signal "
                                         "%s on bus %s" % (sig_name, name))
 
-    def _add_signal(self, attr_name, signame):
-        self._entity._log.debug("Signal name {}".format(signame))
-        setattr(self, attr_name, getattr(self._entity, signame))
+    def _add_signal(self, attr_name, signame, array_idx=None):
+        self._entity._log.debug("Signal name {}, idx {}".format(signame, array_idx))
+        if array_idx is None:
+            setattr(self, attr_name, getattr(self._entity, signame))
+        else:
+            setattr(self, attr_name, getattr(self._entity, signame)[array_idx])
```

Comment:
Probably worth avoiding some duplication here:
```suggestion
        handle = getattr(self._entity, signame)
        if array_idx is None:
            handle = handle[array_idx]
        setattr(self, attr_name, handle)
```